lightning/src/ln/channelmanager.rs

```diff
@@ -2993,26 +3006,69 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					let forward_chan_id = match channel_state.short_to_id.get(&short_chan_id) {
 						Some(chan_id) => chan_id.clone(),
 						None => {
-							failed_forwards.reserve(pending_forwards.len());
 							for forward_info in pending_forwards.drain(..) {
 								match forward_info {
-									HTLCForwardInfo::AddHTLC { prev_short_channel_id, prev_htlc_id, forward_info,
-									                           prev_funding_outpoint } => {
-										let htlc_source = HTLCSource::PreviousHopData(HTLCPreviousHopData {
-											short_channel_id: prev_short_channel_id,
-											outpoint: prev_funding_outpoint,
-											htlc_id: prev_htlc_id,
-											incoming_packet_shared_secret: forward_info.incoming_shared_secret,
-										});
-										failed_forwards.push((htlc_source, forward_info.payment_hash,
-											HTLCFailReason::Reason { failure_code: 0x4000 | 10, data: Vec::new() }
-										));
-									},
+									HTLCForwardInfo::AddHTLC { prev_short_channel_id, prev_htlc_id, forward_info: PendingHTLCInfo {
+										routing, incoming_shared_secret, payment_hash, amt_to_forward, outgoing_cltv_value },
+										prev_funding_outpoint } => {
+											macro_rules! fail_forward {
+												($msg: expr, $err_code: expr, $err_data: expr) => {
+													{
+														log_info!(self.logger, "Failed to accept/forward incoming HTLC: {}", $msg);
+														let htlc_source = HTLCSource::PreviousHopData(HTLCPreviousHopData {
+															short_channel_id: short_chan_id,
+															outpoint: prev_funding_outpoint,
+															htlc_id: prev_htlc_id,
+															incoming_packet_shared_secret: incoming_shared_secret,
+														});
+														failed_forwards.push((htlc_source, payment_hash,
+																HTLCFailReason::Reason { failure_code: $err_code, data: $err_data }
+														));
+														continue;
+													}
+												}
+											}
+											if let PendingHTLCRouting::Forward { onion_packet, .. } = routing {
+												let phantom_secret_res = self.keys_manager.get_node_secret(Recipient::PhantomNode);
+												if phantom_secret_res.is_ok() && fake_scid::is_valid_phantom(&self.fake_scid_rand_bytes, short_chan_id) {
+													let shared_secret = {
+														let mut arr = [0; 32];
+														arr.copy_from_slice(&SharedSecret::new(&onion_packet.public_key.unwrap(), &phantom_secret_res.unwrap())[..]);
+														arr
+													};
+													let next_hop = match onion_utils::decode_next_hop(shared_secret, &onion_packet.hop_data, onion_packet.hmac, payment_hash) {
+														Ok(res) => res,
+														Err(onion_utils::OnionDecodeErr::Malformed { err_msg, err_code }) => {
+															fail_forward!(err_msg, err_code, Vec::new());
+														},
+														Err(onion_utils::OnionDecodeErr::Relay { err_msg, err_code }) => {
+															fail_forward!(err_msg, err_code, Vec::new());
```

Comment:
Oops, we need to handle this and the above case differently. I think the above case may be correct, but here we may end up failing due to a "final node" error but encrypting it as if it came from the second-to-last-hop, which the sender may treat as a bogus failure. Here we should `onion_utils::encrypt_failure_package(onion_utils::build_failure_packet(..))` and then set the forwarding reason to `HTLCFailReason::LightningError`. We also should have a test for "tried to send something bogus, eg wrong amount, to a phantom node, and got back a failure that we decrypted correctly".

Reply:
Sg. I assume we'll want to handle the case 10 lines down (where `construct_recv_pending_htlc_info` errors) similarly as well

Reply:
>  I think the above case may be correct, 

Hmm, I think this case needs to use the phantom shared secret as well? Since we're pretending the phantom is the one creating this error

Reply:
> Hmm, I think this case needs to use the phantom shared secret as well? Since we're pretending the phantom is the one creating this error

Yea, we'll want to encrypt it twice, but we already do the second one, so I think we just need to encrypt here with the phantom shared secret. Tests should show up what's up, though.